game/scripts/vscripts/abilities/sohei/sohei_momentum.lua

```diff
@@ -0,0 +1,410 @@
+
+function modifier_sohei_momentum_knockback:GetEffectAttachType()
+	return PATTACH_ABSORIGIN_FOLLOW
+end
+
+--------------------------------------------------------------------------------
+
+function modifier_sohei_momentum_knockback:CheckState()
+	local state = {
+		[MODIFIER_STATE_NO_UNIT_COLLISION] = true,
+	}
+
+	return state
+end
+
+--------------------------------------------------------------------------------
+
+function modifier_sohei_momentum_knockback:DeclareFunctions()
+	local funcs = {
+		MODIFIER_PROPERTY_OVERRIDE_ANIMATION,
+		MODIFIER_PROPERTY_OVERRIDE_ANIMATION_RATE,
+	}
+
+	return funcs
+end
+
+--------------------------------------------------------------------------------
+
+function modifier_sohei_momentum_knockback:GetOverrideAnimation( event )
+	return ACT_DOTA_FLAIL
+end
+
+--------------------------------------------------------------------------------
+
+if IsServer() then
+	function modifier_sohei_momentum_knockback:OnCreated( event )
+		local unit = self:GetParent()
+		local caster = self:GetCaster()
+
+		local difference = unit:GetAbsOrigin() - caster:GetAbsOrigin()
+
+		-- Movement parameters
+		self.direction = difference:Normalized()
+		self.distance = event.distance
+		self.speed = event.speed
+		self.collision_radius = event.collision_radius
+
+		if self:ApplyHorizontalMotionController() == false then
+			self:Destroy()
+			return
+		end
+	end
+
+--------------------------------------------------------------------------------
+
+	function modifier_sohei_momentum_knockback:OnDestroy()
+		local parent = self:GetParent()
+
+		parent:RemoveHorizontalMotionController( self )
+		ResolveNPCPositions( parent:GetAbsOrigin(), 128 )
+	end
+
+--------------------------------------------------------------------------------
+
+	function modifier_sohei_momentum_knockback:UpdateHorizontalMotion( parent, deltaTime )
+		local caster = self:GetCaster()
+		local parentOrigin = parent:GetAbsOrigin()
+
+		local tickSpeed = self.speed * deltaTime
+		tickSpeed = math.min( tickSpeed, self.distance )
+		local tickOrigin = parentOrigin + ( tickSpeed * self.direction )
+
+		self.distance = self.distance - tickSpeed
+
+		-- If there is at least one target to attack, hit it
+		local targets = FindUnitsInRadius(
+			caster:GetTeamNumber(),
+			tickOrigin,
+			nil,
+			self.collision_radius,
+			DOTA_UNIT_TARGET_TEAM_ENEMY,
+			DOTA_UNIT_TARGET_HERO,
+			bit.bor( DOTA_UNIT_TARGET_FLAG_MAGIC_IMMUNE_ENEMIES, DOTA_UNIT_TARGET_FLAG_NO_INVIS, DOTA_UNIT_TARGET_FLAG_NOT_ATTACK_IMMUNE ),
+			FIND_CLOSEST,
+			false
+		)
+
+		local nonHeroTarget = targets[1]
+		if nonHeroTarget == parent then
+			nonHeroTarget = targets[2]
+		end
+
+		local spell = self:GetAbility()
+
+		if nonHeroTarget then
+			self:SlowAndStun( parent, caster, spell )
+			self:SlowAndStun( nonHeroTarget, caster, spell )
+			self:Destroy()
+		-- why do these mean two different things
+		elseif not GridNav:IsTraversable( tickOrigin ) or GridNav:IsBlocked( tickOrigin ) then
```

Comment:
This made heroes instantly collide with themselves (but creeps don't, only heroes), but i didn't test it with the modifier making them MODIFIER_STATE_NO_UNIT_COLLISION... Definitely worth testing extra, if it doesn't work checking Z diff and IsBlocked should probably work.

Reply:
Works beautifully from my tests. Might've been also caused by the fact that the modifier wasn't a motion controller, so it caused nonsense in terms of blocking.